scalableminds/webknossos
# Read WKW as DatasetArray, no more Memory Mapping

- Removed the dependency webknossos-wrap.
- Took some of its code and copied it into the webknossos codebase (WKWFile,WKWHeader,ResourceBox).
- Renamed some of its properties to match the shard/chunk terminology (previously cube/block)
- Added some methods to the WKWHeader class so that it implements the DatasetHeader interface
- WKWArray now reads WKW data, implements the DatasetArray interface
  - note that AxisOrder is always CXYZ
- Volume annotation downloads/uploads still use some of the original codepaths from WKWFile and WKWHeader, so those are not removed (however, that works without memory mapping)
- Also renamed size to shape where applicable (size just looked too weird, also confusing with flat byte lengths)

### URL of deployed dev instance (used for testing):
- https://wkwdatasetarray.webknossos.xyz

### Steps to test:
- Open wkw dataset,
  - compressed
  - uncompressed
  - multiple shards
  - segmentation
  - different dtypes including uint24, float
- Open other dataset (e.g. zarr)
- Should all load nicely
- Create volume annotation, download + reupload, should show correct content

### TODOs:
- [x] import webknossos-wrap code into wk directly
- [x] DatasetArray
  - [x] header parsing
    - [x] CompressorImpl
      - [x] use streaming lz4 decompression, throw away `expectedUncompressedSizeBytes`?
    - [x] DataType
    - [x] Channels/uint24
    - [x] DatasetShape
  - [x] sharding
  - [x] test color data compressed and uncompressed
  - [x] test segmentation data
  - [x] test with bigger datasets/multiple shards
  - [x] roughly check perf
- [x] how to integrate other usages? (e.g. volume annotation upload/download)
- [x] remove unused code
- [x] do we need to handle the case of missing header.wkw files?
- [x] follow up issues for shard handle caching, ResourceBox, other cleanup

### Issues:
- fixes #7478
- fixes #6818
- follow-up: #7567
- follow-up: #7566

------
